File: Dev/polymorphism.java
/*  Polymorphism, in the context of object-oriented programming (OOP), refers to the ability of different classes to be treated
as instances of the same class through a common interface. This allows objects of different classes to be treated uniformly.
    There are two main types of polymorphism: Compile-time Polymorphism (Method Overloading): This occurs when there are multiple
methods with the same name in a class, but they differ in the number or types of their parameters. During compile-time, 
the correct method to be executed is determined based on the number and types of arguments passed. Example:Calculator 

Run-time Polymorphism (Method Overriding): This occurs when a subclass provides a specific implementation 
of a method that is already defined in its superclass. The method in the subclass overrides the method in 
the superclass. During runtime, the method to be executed is determined by the actual object being referred to.
Example:Animal In the above example, if you have an object of Animal class but it refers to a Dog object,
 calling the sound() method will execute the sound() method from the Dog class, not from the Animal class. This is determined at runtime.
Polymorphism allows for more flexible and modular code, as it enables code reuse, abstraction, and separation of concerns.*/

class Calculator {
    int add(int a, int b) {
        return a + b;
    }

    double add(double a, double b) {
        return a + b;
    }
}


class Animal {
    void sound() {
        System.out.println("Animal makes a sound");
    }
}

class Dog extends Animal {
    void sound() {
        System.out.println("Dog barks");
    }
}

File: Dev/polymorphism.py
 polymorphism in Python 

Python Example:
Method Overloading (Compile-time Polymorphism):
python

class Calculator:
    def add(self, a, b):
        return a + b

    def add(self, a, b, c):
        return a + b + c

calc = Calculator()
print(calc.add(2, 3))      # This will raise an error in Python
print(calc.add(2, 3, 4))   # This will work fine
Python does not directly support method overloading like Java or C++,
 but you can achieve similar functionality by using default parameter values or variable-length argument lists.

Method Overriding (Run-time Polymorphism):
python
Copy code
class Animal:
    def sound(self):
        print("Animal makes a sound")

class Dog(Animal):
    def sound(self):
        print("Dog barks")

animal = Animal()
animal.sound()  # Output: Animal makes a sound

dog = Dog()
dog.sound()     # Output: Dog barks











Ruby Example:
Method Overloading (Compile-time Polymorphism):
ruby
Copy code
class Calculator
  def add(a, b)
    a + b
  end

  def add(a, b, c)
    a + b + c
  end
end

calc = Calculator.new
puts calc.add(2, 3)        # This will raise an error in Ruby
puts calc.add(2, 3, 4)     # This will work fine
Method Overriding (Run-time Polymorphism):
ruby
Copy code
class Animal
  def sound
    puts "Animal makes a sound"
  end
end

class Dog < Animal
  def sound
    puts "Dog barks"
  end
end

animal = Animal.new
animal.sound   # Output: Animal makes a sound

dog = Dog.new
dog.sound      # Output: Dog barks
In both Python and Ruby, method overloading isn't directly supported like in Java,
 but you can achieve similar behavior through different means. However, method overriding
  for achieving run-time polymorphism is supported in both languages.

User
code javascript
ChatGPT
Certainly! Here are the examples of polymorphism in JavaScript:

Method Overloading (Compile-time Polymorphism):
JavaScript does not support method overloading in the same way as languages like Java or C++.
 However, you can achieve similar behavior by checking the number of arguments passed and implementing different logic accordingly.

javascript
Copy code
class Calculator {
    add(a, b, c) {
        if (arguments.length === 2) {
            return a + b;
        } else if (arguments.length === 3) {
            return a + b + c;
        }
    }
}

const calc = new Calculator();
console.log(calc.add(2, 3));        // Output: 5
console.log(calc.add(2, 3, 4));     // Output: 9
Method Overriding (Run-time Polymorphism):
JavaScript supports method overriding for achieving run-time polymorphism.

javascript
Copy code
class Animal {
    sound() {
        console.log("Animal makes a sound");
    }
}

class Dog extends Animal {
    sound() {
        console.log("Dog barks");
    }
}

const animal = new Animal();
animal.sound();   // Output: Animal makes a sound

const dog = new Dog();
dog.sound();      // Output: Dog barks
In JavaScript, you can define methods with the same name in the same class, 
but only the last method definition will take effect, effectively overriding the previous ones. 
This allows you to achieve polymorphic behavior similar to method overriding in other object-oriented languages.




